src_python/fl_client.py

```diff
@@ -0,0 +1,223 @@
+import socket
+import pickle
+import select
+import sys
+import numpy as np
+import pandas as pd
+import logging
+from timeit import default_timer as timer
+import time
+
+arg_names = [
+    'path_weights',
+    'path_nodes',
+    'path_edges',
+    'graph_id',
+    'partition_id',
+    'epochs',
+    'IP',
+    'PORT'
+]
+
+args = dict(zip(arg_names, sys.argv[1:]))
+
+partition_id = args['partition_id']
+
+logging.basicConfig(
+    level=logging.INFO, 
+    format='%(asctime)s : [%(levelname)s]  %(message)s',
+    handlers=[
+        logging.FileHandler(f'client_{partition_id}.log'),
+        logging.StreamHandler(sys.stdout)
+    ]
+)
+class Client:
+
+    def __init__(self, MODEL, graph_params, weights_path, graph_id, partition_id, epochs = 10, IP = socket.gethostname(), PORT = 5000, HEADER_LENGTH = 10):
+
+        self.HEADER_LENGTH =  HEADER_LENGTH
+        self.IP = IP
+        self.PORT = PORT
+
+        self.weights_path = weights_path
+        self.graph_id = graph_id
+        self.partition_id = partition_id
+        self.epochs = epochs
+
+        self.graph_params = graph_params
+
+        self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        connected = False
+        while not connected:
+            try:
+                self.client_socket.connect((IP, PORT))
+            except ConnectionRefusedError:
+                time.sleep(5)
+            else:
+                logging.info('Connected to the server')
+                connected = True
+        
+
+        self.MODEL = MODEL
+        self.STOP_FLAG = False
+        self.rounds = 0
+
+
+    def send_model(self):
+
+        # svae model weights
+        # weights file name : weights_graphid_workerid.npy
+        weights_path = self.weights_path + 'weights_' + self.graph_id + '_' + self.partition_id + ".npy"
+        
+        #np.save(weights_path,self.MODEL.get_weights())
+
+        weights = np.array(self.MODEL.get_weights())
+
+        data = {"CLIENT_ID":self.partition_id,"WEIGHTS":weights,"NUM_EXAMPLES":self.graph_params[0]}
+
+        data = pickle.dumps(data)
+        data = bytes(f"{len(data):<{self.HEADER_LENGTH}}", 'utf-8') + data
+        self.client_socket.sendall(data)
+
+
+    def receive(self):
+        try:
+
+            message_header = self.client_socket.recv(self.HEADER_LENGTH)
+            if not len(message_header):
+                return False
+
+            message_length = int(message_header.decode('utf-8').strip())
+
+            full_msg = b''
+            while True:
+                msg = self.client_socket.recv(message_length)
+
+                full_msg += msg
+
+                if len(full_msg) == message_length:
+                    break
+            
+            data = pickle.loads(full_msg)
+
+            self.STOP_FLAG = data["STOP_FLAG"]
+
+            return data["WEIGHTS"]
+
+        except Exception as e:
+            print(e)
```

Comment:
We need to use proper error logging here rather than using a `print()`. 
Lets use 

```
            logger.error("<Detailed error message>")
            logger.exception(e)
```

Reply:
done

Reply:
Ok